### ISSUE TITLE
Is `lib` a `comptype` and can we target them by name?

### ISSUE THREAD

**Issue by philderbeast** [commercialhaskell/stack#6042]
Could we add `lib` as a  `comptype` in docs section for target syntax?

> * _component_. Instead of referring to an entire package and letting Stack decide which components to build, you select individual components from inside a package. This can be done for more fine-grained control over which test suites to run, or to have a faster compilation cycle. There are multiple ways to refer to a specific component (provided for convenience):
>
> `packagename:comptype:compname` is the most explicit. The available comptypes are `exe`, `test`, and `bench`.
> SOURCE: [Target syntax](https://docs.haskellstack.org/en/stable/build_command/#target-syntax)

There's no mention of `lib` as `comptype` but I can use it.

```
$ stack repl stack:lib
stack> initial-build-steps (lib)
Configuring GHCi with the following packages: stack
GHCi, version 9.2.5: https://www.haskell.org/ghc/  :? for help
[  1 of 117] Compiling Network.HTTP.StackClient
...
[117 of 117] Compiling Stack.Ls
Ok, 117 modules loaded.
ghci>
```

I see it listed as a `packagename:comptype` target:

```
$ stack ide targets
stack:lib
stack:exe:stack
stack:exe:stack-integration-test
stack:test:stack-test
```

The library name is the same as the package name but we can't use that as a `packagename:comptype:compname` target.

```
$ stack repl stack:lib:stack
Error: [S-6948]
TargetParseException ["Directory not found: stack:lib:stack"]
Note that to specify options to be passed to GHCi, use the '--ghci-options' flag.
```

When we support multiple library packages we'll need to be able specify the `compname` of libraries. Before that feature lands, we could allow `compname` for libraries.

**Comment by mpilgrem** [comment]
@philderbeast, it looks to me that that documentation is indeed incomplete, because it does not discuss `packagename:lib`.

Some further notes for my own use:

* Type `Stack.Build.Target.RawTarget` has a data constructor `RTComponent`.

* There is also `Stack.Build.Target.resolveRawTarget`, which accepts (among other things) `RawTarget` and yields a `ResolveResult` (including, among other things, `Maybe NamedComponent`).

* Type `Stack.Types.NamedComponent.NamedComponent` does have a data constructor `CLib`.

**Comment by mpilgrem** [comment]
@philderbeast, I've improved the existing online documentation.

**Comment by wraithm** [comment]
I don't know if this should be a separate issue, but I'm having trouble getting sub-libs into HLS. I suspect it's related to this issue.

**Comment by mpilgrem** [comment]
@theobat, currently, Stack's target syntax `my-package:lib` targets all of the library components of package `my-package`. Given the changes you have already made, do you think it would be difficult to change things so that `my-package:lib` targets only the main library and (eg) `my-package:lib:my-sub-library` targets the named library `my-sub-library`?

**Comment by theobat** [comment]
yes, this is a significant change, mainly because of how the cache system was built in stack, it's package based so building anything less than "all the libraries" means we have to deal with partial values in the cache. I don't remember all the details, but this is somewhat orthogonal to the component based builds : using component based builds, using package based builds we can also build "partially" a package.
Note that for component based builds I made a change in the cache system that is a step in the right direction for this, but it's only a step, dealing with partial cache values is going to be the real thing here.

**Comment by mpilgrem** [comment]
@theobat, many thanks. I'll add 'the cache system' to my list of things I need to understand!
